### PR TITLE
[ldap][2.8] Optional search for primary DN before bind

### PR DESCRIPTION
**Pull request by mschafr**
| Q | A |
| --- | --- |
| Bug fix? | [no] |
| New feature? | [no] |
| BC breaks? | [no] |
| Deprecations? | [no] |
| Tests pass? | [yes] |
| Fixed tickets | [] |
| License | MIT |
| Doc PR | [] |
#### Problem

LDAP bind operation allows only primary Distinguished Name (DN). If this is set to the username (UID), then it is easy to compose the full DN. If on the other hand, the LDAP admin decided to use e.g. the display name as part of the primary DN, then it is not possible to compose the full DN by just having the username and password.
#### Solution

Retrieve the full DN from the LDAP directory using the information we have and then use this for the bind operation.
#### Implementation

Search query is separated by ";" from the rest of the dn.
Example:
Primary DN with a priori unknown CN: `CN=Harry Potter,DC=example,DC=com`
Query DN with known user id: `UID={username};DC=example,DC=com`

### CODE REVIEW

**Comment by fabpot** [comment]
ping @csarrazi

**Comment by xabbuh** [comment]
In any case, the file permission changes must be reverted.

**Comment by csarrazi** [comment]
I'm -1 on this one.

The problem here is that the `bind()` method will break if the admin did not enable anonymous search.

The logic should be moved instead in the `LdapUserProvider`, whose logic is actually to search for users before actually binding to the Ldap server.

**Comment by markussc** [comment]
@csarrazi : Understand your point. But what should we do if e.g. FOSUserBundle is used as user provider and ldap is only needed for authentication? This is exacly my situation..
It seems to me that the "search DN before bind" logic should therefore be placed inside the LdapBindAuthenticationProvider.

**Comment by csarrazi** [comment]
Sorry, my bad, it should indeed be in the `LdapBindAuthenticationProvider`.

**Comment by markussc** [comment]
Just moved it in there with latest commit  b79e663. This now requires that the LdapClient service gets some additional parameters passed in:
`app.ldap:
        class: Symfony\Component\Ldap\LdapClient
        arguments: [ "%ldap_host%", "%ldap_port%", "%ldap_version%", "%ldap_use_ssl%", "%ldap_use_start_tls%", "%ldap_opt_referrals%", "%ldap_base_dn%", "%ldap_search_dn%", "%ldap_search_password%", "%ldap_uid_key%", "%ldap_filter%" ]`

File: src/Symfony/Component/Ldap/LdapClient.php [hunk 1]
```diff
@@ -30,6 +31,7 @@ class LdapClient implements LdapClientInterface
     private $useStartTls;
     private $optReferrals;
     private $connection;
+    private $charmaps;
```
Comment (csarrazi) [inline src/Symfony/Component/Ldap/LdapClient.php]: Remove this field. It is not used.

**Comment by csarrazi** [comment]
Also, this is a new feature. It should not be made from the 2.8 branch, but from the `master` branch.

File: src/Symfony/Component/Ldap/LdapClient.php [hunk 2]
```diff
@@ -41,7 +43,7 @@ class LdapClient implements LdapClientInterface
      * @param bool   $useStartTls
      * @param bool   $optReferrals
      */
-    public function __construct($host = null, $port = 389, $version = 3, $useSsl = false, $useStartTls = false, $optReferrals = false)
+    public function __construct($host = null, $port = 389, $version = 3, $useSsl = false, $useStartTls = false, $optReferrals = false, $ldapBaseDn = null, $ldapSearchDn = null, $ldapSearchPassword = null, $ldapUidKey = null, $ldapFilter = null)
```
Comment (csarrazi) [inline src/Symfony/Component/Ldap/LdapClient.php]: Why did you add these fields? They are not even used...

Comment (markussc) [inline src/Symfony/Component/Ldap/LdapClient.php]: @csarrazi : these fields are passed to the LdapUserProvider inside the LdapBindAuthenticationProvider. This is required in the cases where another than the LdapUserProvider is used for all other purposes than authentication (as described above). Refer to line 87 in LdapBindAuthenticationProvider.php

**Comment by csarrazi** [comment]
Just a little question. Are you trying to use the Ldap component as a standalone library? Or from a Symfony application, using the Security component?

If your intent is to implement this on the Ldap component (standalone), then the code is wrong, as the Ldap component should not rely on the Security component.

Also, I'm against implementing this feature directly in the `Ldap`/`LdapClient` class, as this may break many Ldap configurations. Indeed, one may not have a server authorizing for anonymous bind, simple bind (using UID), or the ability to search entries without password. So in this case, this PR will break the Ldap component. In this case, you first need to `bind()` to the server using the search_dn and search_password, running the search, and then `bind()` once again, with the correct DN.

If you wish to make a search prior to running a Ldap bind(), the only class that needs refactoring is the `LdapAuthenticationProvider` class. However, this means that you will need to duplicate the search logic from the `LdapUserProvider`, which is also wrong.

Also, as mentioned before, this is a *new* feature and, therefore, it should *not* be based (and merged) on the `2.8` branch, but the `master` (3.1) branch instead.

What we can do, though, since `3.1`, is to use an Ldap `Query` (`Symfony\Component\Ldap\Adapter\QueryInterface`) in order to share the Ldap search, and use the same result. Thus, we could reuse the same query for both the user provider and the authentication provider.

**Comment by markussc** [comment]
This is to be used inside a Symfony application, using the Security component.
Actually the implementation is not inside the LdapClient class, but in the LdapBindAuthenticationProvider (lines 82-93). Only the parameters are passed in via the LdapClient, which is not very nice but I don't see an alternative for it.

From my point of view this is compatible both with Ldap configurations using anonymous / non-anonymous bind, since if search_dn and search_password have been specified the search is made using the search-credentials. Otherwise, direct bind with credentials from the login form is made.

Since in the 3.1 branch it seems to be possible to use another aproach (QueryInterface), I think we should maybe close this discussion. Since I am still bound to the 2.8  branch (large project, migration takes a lot of time) I will not be able to implement this feature for 3.1 in the near future. However, since this is really required for most Ldap setups with standard Windows Active Directory servers, I would really appreciate if it could be included in the roadmap.

**Comment by csarrazi** [comment]
@markussc I do agree that it should be included in the roadmap.

One thing, though. Currently, we already are able to authenticate against Active Directory servers. For that, we simply override the dn_string, to be as follows: `{username}@MYDOMAIN`, which works perfectly.

**Comment by markussc** [comment]
I do not know how the "standard" DN scheme looks when configuring an AD server (I never did that myself). I just found that the example configuration by Microsofts uses exact the scheme which is used in the customer system I am supposed to bind to (refer to https://msdn.microsoft.com/en-us/library/windows/desktop/aa366101%28v=vs.85%29.aspx) : DN="CN={{displayName}}, ..."  Unfortunately, this makes it impossible to use the {username}@MYDOMAIN scheme. My assumption (and the answer from the AD responsible regarding the composition of DN names ) is, that by default, Microsoft uses the display name in the DN.
For my own test-system, I setup a openLdap server, where I used the username in the DN. There, your suggestion works perfectly.

**Comment by csarrazi** [comment]
Did you try using `UID={username}` in the `dn_string`?

**Comment by csarrazi** [comment]
Regardless, this is indeed a new feature. We could implement the bind using different strategies:

* Simple bind
* Search bind
* etc.

@fabpot Should we create a new issue for this, and close this PR?

**Comment by fabpot** [comment]
Sure, I'm closing this PR and I let you create the issue.

**Comment by markussc** [comment]
@csarrazi : yes, tried all the combinations. The point is, that for a bind, the DN used at creation of the entry MUST be used. Windows AD denies all alternative DNs. Since they used the "displayName" in the DN, I am stuck...
Thanks for taking this feature to the agenda! I'll be happy to assist, at least with testing!

**Comment by csarrazi** [comment]
What about `samAccountName={username}`?

It is weird that you still have an authentication issue, as from the official Microsoft documentation, AD tries to find the user using different strategies, one after the other: https://msdn.microsoft.com/en-us/library/cc223499.aspx

Regardless, this will be added in a future iteration.

**Comment by markussc** [comment]
This is really really strange, but you made my day!
The following patterns work:
`<domain>\{usenrame}` and `<email-address>@FQDN` . I can use the first one.
With the `sAMAccountName`, it didn't work...
As said before, the feature will be a good add-on for the future anyway! Thanks a lot for helping!

**Comment by csarrazi** [comment]
You're welcome! I figured it was really strange, that nothing worked! :smiley:
